batch/batch/front_end/query/query_v2.py

```diff
@@ -238,6 +238,18 @@ def parse_batch_jobs_query_v2(batch_id: int, q: str, last_job_id: Optional[int])
     where_conditions = ['(jobs.batch_id = %s AND batch_updates.committed)']
     where_args = [batch_id]
 
+    if recursive:
+        jg_cond = """
+((jobs.batch_id, jobs.job_group_id) IN
+ (SELECT batch_id, job_group_id FROM job_group_self_and_ancestors
+  WHERE ancestor_id = %s))
```

Comment:
Original:

```
mysql> EXPLAIN SELECT jobs.*, batches.user, batches.billing_project, batches.format_version, job_attributes.value AS name, cost_t.cost,
    ->   cost_t.cost_breakdown
    -> FROM jobs
    -> INNER JOIN batches ON jobs.batch_id = batches.id
    -> INNER JOIN batch_updates ON jobs.batch_id = batch_updates.batch_id AND jobs.update_id = batch_updates.update_id
    -> LEFT JOIN job_attributes
    ->   ON jobs.batch_id = job_attributes.batch_id AND
    ->     jobs.job_id = job_attributes.job_id AND
    ->     job_attributes.`key` = 'name'
    -> LEFT JOIN LATERAL (
    -> SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    -> FROM (SELECT aggregated_job_resources_v3.batch_id, aggregated_job_resources_v3.job_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    ->   FROM aggregated_job_resources_v3
    ->   WHERE aggregated_job_resources_v3.batch_id = jobs.batch_id AND aggregated_job_resources_v3.job_id = jobs.job_id
    ->   GROUP BY aggregated_job_resources_v3.batch_id, aggregated_job_resources_v3.job_id, aggregated_job_resources_v3.resource_id
    -> ) AS usage_t
    -> LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> GROUP BY usage_t.batch_id, usage_t.job_id
    -> ) AS cost_t ON TRUE
    -> WHERE (jobs.batch_id = 1110000 AND batch_updates.committed) AND ((jobs.batch_id, jobs.job_group_id) IN
    ->  (SELECT batch_id, job_group_id FROM job_group_self_and_ancestors
    ->   WHERE ancestor_id = 0));
+----+-------------------+------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+---------------------------------------+------+----------+-----------------------------------------+
| id | select_type       | table                        | partitions | type   | possible_keys                                                                                                                                                                                                                                                         | key                                            | key_len | ref                                   | rows | filtered | Extra                                   |
+----+-------------------+------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+---------------------------------------+------+----------+-----------------------------------------+
|  1 | PRIMARY           | batches                      | NULL       | const  | PRIMARY                                                                                                                                                                                                                                                               | PRIMARY                                        | 8       | const                                 |    1 |   100.00 | NULL                                    |
|  1 | PRIMARY           | job_group_self_and_ancestors | NULL       | ref    | PRIMARY,job_group_self_and_ancestors_ancestor_id_level,job_group_self_and_ancestors_job_group_id_level                                                                                                                                                                | job_group_self_and_ancestors_ancestor_id_level | 12      | const,const                           |    1 |   100.00 | Using index                             |
|  1 | PRIMARY           | batch_updates                | NULL       | ref    | PRIMARY,batch_id,batch_updates_committed,batch_updates_start_job_id                                                                                                                                                                                                   | PRIMARY                                        | 8       | const                                 |    1 |    90.00 | Using where                             |
|  1 | PRIMARY           | jobs                         | NULL       | ref    | PRIMARY,jobs_batch_id_state_always_run_cancelled,jobs_batch_id_state_always_run_inst_coll_cancelled,jobs_batch_id_update_id,jobs_batch_id_always_run_n_regions_regions_bits_rep_job_id,jobs_batch_id_ic_state_ar_n_regions_bits_rep_job_id,jobs_batch_id_job_group_id | PRIMARY                                        | 8       | const                                 |   16 |     1.00 | Using where; Rematerialize (<derived2>) |
|  1 | PRIMARY           | job_attributes               | NULL       | eq_ref | PRIMARY,job_attributes_key_value,job_attributes_batch_id_key_value,job_attributes_value                                                                                                                                                                               | PRIMARY                                        | 314     | const,batch.jobs.job_id,const         |    1 |   100.00 | NULL                                    |
|  1 | PRIMARY           | <derived2>                   | NULL       | ALL    | NULL                                                                                                                                                                                                                                                                  | NULL                                           | NULL    | NULL                                  |    5 |   100.00 | Using where                             |
|  2 | DEPENDENT DERIVED | <derived3>                   | NULL       | ALL    | NULL                                                                                                                                                                                                                                                                  | NULL                                           | NULL    | NULL                                  |    5 |   100.00 | Using filesort                          |
|  2 | DEPENDENT DERIVED | resources                    | NULL       | eq_ref | resource_id                                                                                                                                                                                                                                                           | resource_id                                    | 4       | usage_t.resource_id                   |    1 |   100.00 | NULL                                    |
|  3 | DEPENDENT DERIVED | aggregated_job_resources_v3  | NULL       | ref    | PRIMARY,resource_id                                                                                                                                                                                                                                                   | PRIMARY                                        | 12      | batch.jobs.batch_id,batch.jobs.job_id |    5 |   100.00 | NULL                                    |
+----+-------------------+------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+---------------------------------------+------+----------+-----------------------------------------+
9 rows in set, 3 warnings (0.07 sec)
```

Your proposed change:
```
mysql> EXPLAIN SELECT jobs.*, batches.user, batches.billing_project, batches.format_version, job_attributes.value AS name, cost_t.cost,
    ->   cost_t.cost_breakdown
    -> FROM jobs
    -> INNER JOIN batches ON jobs.batch_id = batches.id
    -> INNER JOIN batch_updates ON jobs.batch_id = batch_updates.batch_id AND jobs.update_id = batch_updates.update_id
    -> LEFT JOIN job_attributes
    ->   ON jobs.batch_id = job_attributes.batch_id AND
    ->     jobs.job_id = job_attributes.job_id AND
    ->     job_attributes.`key` = 'name'
    -> LEFT JOIN LATERAL (
    -> SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    -> FROM (SELECT aggregated_job_resources_v3.batch_id, aggregated_job_resources_v3.job_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    ->   FROM aggregated_job_resources_v3
    ->   WHERE aggregated_job_resources_v3.batch_id = jobs.batch_id AND aggregated_job_resources_v3.job_id = jobs.job_id
    ->   GROUP BY aggregated_job_resources_v3.batch_id, aggregated_job_resources_v3.job_id, aggregated_job_resources_v3.resource_id
    -> ) AS usage_t
    -> LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> GROUP BY usage_t.batch_id, usage_t.job_id
    -> ) AS cost_t ON TRUE
    -> WHERE (jobs.batch_id = 1110000 AND batch_updates.committed) AND ((jobs.batch_id, jobs.job_group_id) IN
    ->  (SELECT batch_id, job_group_id FROM job_group_self_and_ancestors
    ->   WHERE batch_id = 1110000 AND ancestor_id = 0));
+----+-------------------+------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+---------------------------------------+------+----------+-----------------------------------------+
| id | select_type       | table                        | partitions | type   | possible_keys                                                                                                                                                                                                                                                         | key                                            | key_len | ref                                   | rows | filtered | Extra                                   |
+----+-------------------+------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+---------------------------------------+------+----------+-----------------------------------------+
|  1 | PRIMARY           | batches                      | NULL       | const  | PRIMARY                                                                                                                                                                                                                                                               | PRIMARY                                        | 8       | const                                 |    1 |   100.00 | NULL                                    |
|  1 | PRIMARY           | job_group_self_and_ancestors | NULL       | ref    | PRIMARY,job_group_self_and_ancestors_ancestor_id_level,job_group_self_and_ancestors_job_group_id_level                                                                                                                                                                | job_group_self_and_ancestors_ancestor_id_level | 12      | const,const                           |    1 |   100.00 | Using index                             |
|  1 | PRIMARY           | batch_updates                | NULL       | ref    | PRIMARY,batch_id,batch_updates_committed,batch_updates_start_job_id                                                                                                                                                                                                   | PRIMARY                                        | 8       | const                                 |    1 |    90.00 | Using where                             |
|  1 | PRIMARY           | jobs                         | NULL       | ref    | PRIMARY,jobs_batch_id_state_always_run_cancelled,jobs_batch_id_state_always_run_inst_coll_cancelled,jobs_batch_id_update_id,jobs_batch_id_always_run_n_regions_regions_bits_rep_job_id,jobs_batch_id_ic_state_ar_n_regions_bits_rep_job_id,jobs_batch_id_job_group_id | PRIMARY                                        | 8       | const                                 |   16 |     1.00 | Using where; Rematerialize (<derived2>) |
|  1 | PRIMARY           | job_attributes               | NULL       | eq_ref | PRIMARY,job_attributes_key_value,job_attributes_batch_id_key_value,job_attributes_value                                                                                                                                                                               | PRIMARY                                        | 314     | const,batch.jobs.job_id,const         |    1 |   100.00 | NULL                                    |
|  1 | PRIMARY           | <derived2>                   | NULL       | ALL    | NULL                                                                                                                                                                                                                                                                  | NULL                                           | NULL    | NULL                                  |    5 |   100.00 | Using where                             |
|  2 | DEPENDENT DERIVED | <derived3>                   | NULL       | ALL    | NULL                                                                                                                                                                                                                                                                  | NULL                                           | NULL    | NULL                                  |    5 |   100.00 | Using filesort                          |
|  2 | DEPENDENT DERIVED | resources                    | NULL       | eq_ref | resource_id                                                                                                                                                                                                                                                           | resource_id                                    | 4       | usage_t.resource_id                   |    1 |   100.00 | NULL                                    |
|  3 | DEPENDENT DERIVED | aggregated_job_resources_v3  | NULL       | ref    | PRIMARY,resource_id                                                                                                                                                                                                                                                   | PRIMARY                                        | 12      | batch.jobs.batch_id,batch.jobs.job_id |    5 |   100.00 | NULL                                    |
+----+-------------------+------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+---------------------------------------+------+----------+-----------------------------------------+
9 rows in set, 3 warnings (0.01 sec)
```

Reply:
Looks identical to me. Let me know whether you still prefer adding the batch_id to the where clause.

Reply:
Yeah, it seems like MySQL is able to clean up the missing batch_id regardless, but I'm more curious about the EXISTS vs the IN. I used your namespace's tables which appear to use job_groups_parents instead of the self_and_ancestors terminology. Also seems to be the same.

I'm ambivalent on EXISTS versus IN but I really think we should include the batch_id in the WHERE clause. I do not like relying on MySQL to realize that the only possible batch_id is batches.id and then using that to filter the sub-query. I'd worry that a seemingly unrelated change could prevent MySQL's query optimizer from doing that. That'd be super annoying to debug :)

---

```sql
mysql> EXPLAIN SELECT jobs.*, batches.user, batches.billing_project, batches.format_version, job_attributes.value AS name, cost_t.cost,
    ->   cost_t.cost_breakdown
    -> FROM jobs
    -> INNER JOIN batches ON jobs.batch_id = batches.id
    -> INNER JOIN batch_updates ON jobs.batch_id = batch_updates.batch_id AND jobs.update_id = batch_updates.update_id
    -> LEFT JOIN job_attributes
    ->   ON jobs.batch_id = job_attributes.batch_id AND
    ->     jobs.job_id = job_attributes.job_id AND
    ->     job_attributes.`key` = 'name'
    -> LEFT JOIN LATERAL (
    -> SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    -> FROM (SELECT aggregated_job_resources_v3.batch_id, aggregated_job_resources_v3.job_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    ->   FROM aggregated_job_resources_v3
    ->   WHERE aggregated_job_resources_v3.batch_id = jobs.batch_id AND aggregated_job_resources_v3.job_id = jobs.job_id
    ->   GROUP BY aggregated_job_resources_v3.batch_id, aggregated_job_resources_v3.job_id, aggregated_job_resources_v3.resource_id
    -> ) AS usage_t
    -> LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> GROUP BY usage_t.batch_id, usage_t.job_id
    -> ) AS cost_t ON TRUE
    -> WHERE (jobs.batch_id = 1 AND batch_updates.committed) AND (EXISTS (SELECT TRUE
    ->   FROM job_group_parents
    ->   WHERE batch_id = 1 AND parent_id = 0 AND job_group_id = jobs.job_group_id));
+----+-------------------+-----------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------+---------------------------------------------+------+----------+-----------------------------------------+
| id | select_type       | table                       | partitions | type   | possible_keys                                                                                                                                                                                                                                                         | key                        | key_len | ref                                         | rows | filtered | Extra                                   |
+----+-------------------+-----------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------+---------------------------------------------+------+----------+-----------------------------------------+
|  1 | PRIMARY           | batches                     | NULL       | const  | PRIMARY                                                                                                                                                                                                                                                               | PRIMARY                    | 8       | const                                       |    1 |   100.00 | NULL                                    |
|  1 | PRIMARY           | batch_updates               | NULL       | ref    | PRIMARY,batch_id,batch_updates_committed,batch_updates_start_job_id                                                                                                                                                                                                   | PRIMARY                    | 8       | const                                       |    1 |    88.89 | Using where                             |
|  1 | PRIMARY           | job_group_parents           | NULL       | ref    | PRIMARY,batch_id,job_group_parents_parent_id,job_group_parents_level                                                                                                                                                                                                  | batch_id                   | 12      | const,const                                 |    1 |   100.00 | Using index                             |
|  1 | PRIMARY           | jobs                        | NULL       | ref    | PRIMARY,jobs_batch_id_state_always_run_cancelled,jobs_batch_id_state_always_run_inst_coll_cancelled,jobs_batch_id_update_id,jobs_batch_id_always_run_n_regions_regions_bits_rep_job_id,jobs_batch_id_ic_state_ar_n_regions_bits_rep_job_id,jobs_batch_id_job_group_id | jobs_batch_id_job_group_id | 12      | const,jigold.job_group_parents.job_group_id |    1 |    10.00 | Using where; Rematerialize (<derived2>) |
|  1 | PRIMARY           | job_attributes              | NULL       | eq_ref | PRIMARY,job_attributes_key_value                                                                                                                                                                                                                                      | PRIMARY                    | 314     | const,jigold.jobs.job_id,const              |    1 |   100.00 | NULL                                    |
|  1 | PRIMARY           | <derived2>                  | NULL       | ALL    | NULL                                                                                                                                                                                                                                                                  | NULL                       | NULL    | NULL                                        |    6 |   100.00 | Using where                             |
|  2 | DEPENDENT DERIVED | <derived3>                  | NULL       | ALL    | NULL                                                                                                                                                                                                                                                                  | NULL                       | NULL    | NULL                                        |    6 |   100.00 | Using filesort                          |
|  2 | DEPENDENT DERIVED | resources                   | NULL       | eq_ref | resource_id                                                                                                                                                                                                                                                           | resource_id                | 4       | usage_t.resource_id                         |    1 |   100.00 | NULL                                    |
|  3 | DEPENDENT DERIVED | aggregated_job_resources_v3 | NULL       | ref    | PRIMARY,resource_id                                                                                                                                                                                                                                                   | PRIMARY                    | 12      | jigold.jobs.batch_id,jigold.jobs.job_id     |    6 |   100.00 | NULL                                    |
+----+-------------------+-----------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------+---------------------------------------------+------+----------+-----------------------------------------+
```